README.rst
canonicalwebteam.gsa: Python GSA client
===================================

.. image:: https://travis-ci.org/canonicalwebteam/gsa.svg?branch=master
   :alt: build status
   :target: https://travis-ci.org/canonicalweeam/gsa

A client library for the `Google Search Appliance <https://enterprise.google.com/search/products/gsa.html>`_, to make retrieving search results in Python easier.

Installation
------------

This module is in PyPi as :code:`canonicalwebteam.gsa`. You should be able to install it simply with:

.. code:: bash

    pip install canonicalwebteam.gsa

GSAClient
---------

This is a basic client for querying a Google Search Appliance.

Making queries
~~~~~~~~~~~~~~

You can query the GSA using the :code:`search` method.

.. code:: python

    search_client = GSAClient(base_url="http://gsa.example.com/search")

    first_ten_results = search_client.search("hello world")

    first_thirty_results = search_client.search("hello world", num=30)

    results_twenty_to_forty = search_client.search(
      "hello world", start=20, num=20
    )

This will set the `q <https://www.google.com/support/enterprise/static/gsa/docs/admin/72/gsa_doc_set/xml_reference/request_format.html#1089652>`_,
`start <https://www.google.com/support/enterprise/static/gsa/docs/admin/72/gsa_doc_set/xml_reference/request_format.html#1076971>`_ (default: 0) and
`num <https://www.google.com/support/enterprise/static/gsa/docs/admin/72/gsa_doc_set/xml_reference/request_format.html#1076882>`_ (default: 10) and
`lr <https://www.google.com/support/enterprise/static/gsa/docs/admin/72/gsa_doc_set/xml_reference/request_format.html#1076879>`_ (default: '') parameters.
No other `search parameters <https://www.google.com/support/enterprise/static/gsa/docs/admin/72/gsa_doc_set/xml_reference/request_format.html#1086546>`_,
will be provided, so they will all fall back to their defaults.

The returned results object will attempt to map each of the GSA's
`standard result XML tags <https://www.google.com/support/enterprise/static/gsa/docs/admin/70/gsa_doc_set/xml_reference/results_format.html#1078461>`_
into a more readable format:

.. code:: python

    {
        'estimated_total_results': int,  # "M": GSA's estimate, see below
        'document_filtering': bool,      # "FI": Is filtering enabled?
        'next_url': str,                 # "NU": GSA URL for querying the next set of results, if available
        'previous_url': str,             # "PU": Ditto for previous set of results
        'items': [
            {
                'index': int,            # "R[N]": The number of this result in the index of all results
                'url': str,              # "U": The URL of the resulting page
                'encoded_url': str,      # "UE": The above URL, encoded
                'title': str,            # "T": The page title
                'relevancy': int,        # "RK": How relevant is this result to the query? From 0 to 10
                'appliance_id': str,     # "ENT_SOURCE": The serial number of the GSA
                'summary': str,          # "S": Summary text for this result
                'language': str,         # "LANG": The language of the page
                'details': {}            # "FS": Name:value pairs of any extra info
                'link_supported': bool,  # "L": “link:” special query term is supported,
                'cache': {               # "C": Dictionary, or "None" if cache is not available
                    'size': str,         # "C[SZ]": Human readable size of cached page
                    'cache_id': str,     # "C[CID]": ID of document in GSA's cache
                    'encoding': str      # "C[ENC]": The text encoding of the cached page
                }
            },
            ...
        ]
    }

Filtering by domain or language
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can filter your search results by specifying specific domains or a
`specific language <https://www.google.com/support/enterprise/static/gsa/docs/admin/72/gsa_doc_set/xml_reference/request_format.html#1077439>`_.

.. code:: python

    english_results = search_client.search("hello world", language="lang_en")
    non_english_results = search_client.search("hello world", language="-lang_en")
    domain_specific_results = search_client.search(
        "hello world",
        domains=["site1.example.com", "site2.example.com"]
    )

*NB:* If no search results are found with the specified ``language``, the GSA will fall back to returning any results it finds in all languages.

Getting accurate totals
~~~~~~~~~~~~~~~~~~~~~~~

At the time of writing, the Google Search Appliance will return an "estimate" of
the total number of results with each query, but this estimate is usually wildly
inaccurate, sometimes out by more than a factor of 10! This is true even with
`rc <https://www.google.com/support/enterprise/static/gsa/docs/admin/72/gsa_doc_set/xml_reference/request_format.html#1076946>`_
enabled.

With the :code:`total_results` method, the client will attempt to request results
990 - 1000. This will usually result in the GSA returning the last page of
results, which allows us to find the actual total number of results.

.. code:: python

    total = search_client.total_results("hello world", domains=[], language='')

Django view
-----------

To simplify usage of the GSA client with Django, a Django view is included
with this module.

Usage
~~~~~

At the minimum, need to provide the :code:`SEARCH_SERVER_URL` setting to tell the view
where to find the GSA:

.. code:: python

    # settings.py
    SEARCH_SERVER_URL = 'http://gsa.example.com/search'  # Required: GSA location
    SEARCH_DOMAINS = ['site1.example.com']               # Optional: By default, limit results to this set of domains
    SEARCH_LANGUAGE = 'lang_zh-CN'                       # Optional: By default, limit results to this language

    # urls.py
    from canonicalwebteam.gsa.views import SearchView
    urlpatterns += [url(r'^search/?$', SearchView.as_view(template_name="search.html"))]

This view will then be available to be queried:

- :code:`example.com/search?q=my+search+term`
- :code:`example.com/search?q=my+search+term&domain=example.com&domain=something.example.com`  (overrides :code:`SEARCH_DOMAINS`)
- :code:`example.com/search?q=my+search+term&language=-lang_zh-CN`  (exclude results in Chinese, overrides :code:`SEARCH_LANGUAGE`)

After retrieving search results, the view will pass the context object to the specified :code:`template_name` (in this case :code:`search.html`).

The context object will be structured as follows:

.. code:: python

    {
        'query': str,       # The value of the `q` parameters passed to the view
        'limit': int,       # The value of the `limit` parameter, or the default of 10
        'offset': int,      # The value of the `offset` parameter, or the default of 0
        'error': None|str,  # None, or a description of the error if one occurred
        'results': {
            'items': [],    # The list of items as returned from the GSAClient (see above)
            'total': int,   # The exact total number of results available
            'start': int,   # The index of the first result in the set
            'end': int,     # The index of the last result in the set
            'next_offset': int|None,      # The offset for the next page of results, if available
            'previous_offset': int|None,  # The offset for the previous page of results, if available
            'last_page_offset': int,      # The offset for the last page of results
            'last_page': int,             # The final page number (calculated from "limit" and "total")
            'current_page': int,          # The current page number (calculated from "limit" and "end")
            'penultimate_page': int       # The second-to-last page
    }


CHANGES.txt
v0.1.0, 2017-01-25 -- Initial release
v0.1.1, 2017-01-25 -- Support filtering by domain in GSAClient
v1.0.0, 2017-01-26 -- Support filtering by language; restructure data objects
v1.0.1, 2017-01-27 -- Fix domain filtering
v1.0.2, 2017-01-27 -- Be more permissive with dependency versioning
v1.0.3, 2017-01-31 -- Timouts for communicating with search server
v1.0.4, 2017-01-31 -- Catch timeout errors properly in the view
v1.0.5, 2017-02-21 -- Pass the "domains" info through to the template
v1.1.0, 2017-05-22 -- Python3 compatibility
v1.2.0, 2018-01-10 -- Use GSA estimated total for amount of results.
v2.0.0, 2018-04-20 -- Rename to canonicalwebteam.gsa
v2.0.1, 2018-05-08 -- Hotfix: Total results parse error


canonicalwebteam/gsa/views.py
# Core modules
import math
try:
    from urllib.error import URLError
except:
    from urllib2 import URLError
import requests

# Third party modules
from django.conf import settings
from django.views.generic import TemplateView

# Local modules
from . import GSAClient


class SearchView(TemplateView):
    '''
    A Django view for querying search results from the Google Search Appliance
    and passing them to a Django template.

    Requests should be formatted: <url>?q=<string>&offset=<num>&limit=<num>

    I've used "offset" and "limit" for pagination,
    following the "Web API Design" standard:
    https://pages.apigee.com/web-api-design-ebook.html

    This expects to find the search server URL
    in the SEARCH_SERVER_URL Django setting.

    E.g.: SEARCH_SERVER_URL = 'http://my-server.com/search'
    '''

    def get_context_data(self, **kwargs):
        """
        Extend TemplateView.get_context_data to parse query parameters
        and return search results from the Google Search Appliance (GSA)

        E.g.: http://example.com/search?q=juju&limit=10&offset=10

        Query parameters:
        - q: the search query to be passed to the GSA
        - limit: number of results to return, "page size" (default: 10)
        - offset: where to start results at (default: 0)
        """

        search_server_url = settings.SEARCH_SERVER_URL
        search_client = GSAClient(search_server_url)

        query = self.request.GET.get('q', '').encode('utf-8')
        domains = (
            self.request.GET.getlist('domain') or
            getattr(settings, "SEARCH_DOMAINS", [])
        )
        language = (
            self.request.GET.get('language') or
            getattr(settings, "SEARCH_LANGUAGE", '')
        )
        timeout = (
            self.request.GET.get('language') or
            getattr(settings, "SEARCH_TIMEOUT", 30)
        )
        limit = int(self.request.GET.get('limit', '10'))
        offset = int(self.request.GET.get('offset', '0'))
        results = {}
        error = None

        if query:
            try:
                server_results = search_client.search(
                    query,
                    start=offset,
                    num=limit,
                    domains=domains,
                    language=language,
                    timeout=timeout
                )
                items = server_results['items']

                total = int(server_results['estimated_total_results'])

                results = {
                    'items': items,
                    'total': total,
                }

                if total > 0:
                    start = items[0]['index']
                    end = items[-1]['index']
                    last_page = int(math.ceil(float(total) / limit))
                    penultimate_page = last_page - 1
                    current_page = int(math.ceil(float(end) / limit))
                    last_page_offset = limit * penultimate_page
                    next_offset = offset + limit
                    previous_offset = offset - limit
                    if next_offset >= total:
                        next_offset = None
                    if previous_offset < 0:
                        previous_offset = None

                    results.update({
                        'start': start,
                        'end': end,
                        'next_offset': next_offset,
                        'previous_offset': previous_offset,
                        'last_page_offset': last_page_offset,
                        'last_page': last_page,
                        'penultimate_page': penultimate_page,
                        'current_page': current_page
                    })

                    if offset + limit < total:
                        results['next_offset'] = offset + limit
                    if offset - limit >= 0:
                        results['previous_offset'] = offset - limit

            except URLError:
                error = 'URL error'
            except requests.ConnectionError:
                error = 'connection error'
            except requests.Timeout:
                error = 'timeout error'
            except requests.RequestException:
                error = 'general request error'

        # Import context from parent
        template_context = super(SearchView, self).get_context_data(**kwargs)

        template_context['results'] = results
        template_context['query'] = query
        template_context['limit'] = limit
        template_context['offset'] = offset
        template_context['domains'] = domains
        template_context['error'] = error

        return template_context


ubuntudesign/gsa/__init__.py
# Core modules
try:
    from urllib.parse import urlencode
except:
    from urllib import urlencode

# Third party modules
import requests
from lxml import etree


def xml_text(root_element, child_tag):
    if root_element.xpath(child_tag):
        return root_element.xpath(child_tag)[0].text


class GSAClient:
    """
    Query the Google Search Appliance and return results
    as a python dictionary.

    Usage:

    search_client = GSAClient(base_url="http://gsa.example.com/search")
    results = search_client.search("hello world")
    total = search_client.total_results("hello world")
    """

    def __init__(self, base_url):
        self.base_url = base_url

    def search(
        self, query,
        start=0, num=10, domains=[], language="", timeout=30
    ):
        """
        Query the GSA to get response in XML format
        which it will then parse into a dictionary.
        """

        query_string = query.decode()

        # Filter by domains, if specified
        if domains:
            domain_filters = ['site:' + domain for domain in domains]
            query_string += ' ( ' + " | ".join(domain_filters) + ' )'

        # Build the GSA URL
        query_parameters = urlencode({
            'q': query_string,
            'num': str(num),
            'start': str(start),
            'lr': language
        })
        search_url = self.base_url + '?' + query_parameters

        response = requests.get(search_url, timeout=timeout)

        xml_tree = etree.fromstring(response.content)

        # We're now going to parse the XML items into a hopefully
        # more meaningful dictionary object.
        # To understand the layout of the XML document, see here:
        # https://www.google.com/support/enterprise/static/gsa/docs/admin/70/gsa_doc_set/xml_reference/results_format.html

        results = {
            'estimated_total_results': xml_text(xml_tree, '/GSP/RES/M'),
            'document_filtering': bool(xml_text(xml_tree, '/GSP/RES/FI')),
            'next_url': xml_text(xml_tree, '/GSP/RES/NB/NU'),
            'previous_url': xml_text(xml_tree, '/GSP/RES/NB/PU'),
            'items': []
        }

        item_elements = xml_tree.xpath('/GSP/RES/R')

        for item_element in item_elements:
            item = {
                'index': int(item_element.attrib.get('N')),
                'url': xml_text(item_element, 'U'),
                'encoded_url': xml_text(item_element, 'UE'),
                'title': xml_text(item_element, 'T'),
                'relevancy': int(xml_text(item_element, 'RK')),
                'appliance_id': xml_text(item_element, 'ENT_SOURCE'),
                'summary': xml_text(item_element, 'S'),
                'language': xml_text(item_element, 'LANG'),
                'details': {},
                'link_supported': bool(item_element.xpath('HAS/L')),
                'cache': None
            }

            detail_elements = item_element.xpath('FS')

            for detail in detail_elements:
                item['details'][detail.attrib['NAME']] = detail.attrib['VALUE']

            cache_elements = item_element.xpath('HAS/C')

            if cache_elements:
                item['cache'] = {
                    'size': cache_elements[0].attrib.get('SZ'),
                    'cache_id': cache_elements[0].attrib.get('CID'),
                    'encoding': cache_elements[0].attrib.get('ENC')
                }

            results['items'].append(item)

        return results
